packages/@aws-cdk/aws-lambda/lib/event-source-mapping.ts

```diff
@@ -74,6 +74,29 @@ export interface SourceAccessConfiguration {
   readonly uri: string
 }
 
+/**
+ * Filter rules applied to a key or path. A rule consists of an operator and/or arguments.
+ *
+ * @see https://docs.aws.amazon.com/lambda/latest/dg/invocation-eventfiltering.html
+ */
+export type FilterPatternRule = (
```

Comment:
I think we can do better than this honestly. The CDK pattern here is usually a `FilterPatternRule` class with static methods:

```ts
export class FilterPatternRule {
	public static serializedJson(jsonString: string) {}
	public static excludes()
	public static numeric()
	public static exists()
	public static prefix()
}
```

and go from there. For example, `numeric()` might look like this:

```ts
export interface Numeric {
	operator: '<' | '<=' | '=' | '>=' | '>';
	number: number;
}
...
public static numeric(numerics: Numeric[]): {
	const numericList = [];
	numerics.forEach((n) => {
		numericList.push(n.operator);
		numericList.push(n.number);
	});
	const rule = { "numeric": numericList };
	return FilterPatternRule(rule);
}
```

Just a general idea as I haven't looked into the syntax we're trying to model much, but hopefully you get a sense of how I'd like it to look versus what you currently suggest.

Reply:
I definitely do have the time to iterate on this. I am also convinced we have to discuss the design thoroughly before we finalize on an implementation.

I see, I already guessed wrapping it in a class is a better pattern for CDK, but wanted to first give it a try with a simpler approach.

I think this might take it on a more complex level though, as it might create a much more verbose and maybe convoluted syntax compared to the underlying JSON if we have to wrap the `FilterPattern` as well. But let's give it a try.


Let's look at a more complex example of the rules with the `FilterPatternRule` class and static methods as I understand it so far and how it could be used in TypeScript:
```ts
new sources.DynamoEventSource(table, {
  filterPatterns: [
    {
      dynamodb: {
        field: [
          FilterPatternRule.exists(),
          FilterPatternRule.prefix("arn::"),
        ],
      },
      source: [ FilterPatternRule.prefix("arn::") ],
    },
  ],
})
```

We need to support the following scenarios:

1. Simple rule matching a root-level field (here `dynamodb`):
    ```ts
    new sources.DynamoEventSource(table, {
      filterPatterns: [
        { dynamodb: [ FilterPatternRule.exists() ] },
      ],
    })
    ```
2. Composed rule matching a root-level field:
    ```ts
    new sources.DynamoEventSource(table, {
      filterPatterns: [
        {
          dynamodb: [
            FilterPatternRule.exists(),
            FilterPatternRule.prefix("test-"),
          ],
        },
      ],
    })
    ```
3. rule matching multiple root-level fields
     ```ts
    new sources.DynamoEventSource(table, {
      filterPatterns: [
        {
          dynamodb: [ FilterPatternRule.exists() ],
          source: [ FilterPatternRule.exists() ],
        },
      ],
    })
    ```
4. rule matching nested fields
    ```ts
    new sources.DynamoEventSource(table, {
      filterPatterns: [
        {
          dynamodb: {
            field: [ FilterPatternRule.exists() ],
          },
        },
      ],
    })
    ```
    This case specifically might cause some problems for jsii, as the type of `dynamodb` can be an `object` or a `FilterPatternRule`.

Looking at the `FilterPatternRule` class, I think we need the following static methods, which should be easy to implement:
1. `null`, just matching the event field value with null
2. `match`, which takes a string argument to compare the field value with
3. `numeric`, which takes the Numeric interface proposed by you (which I really like)
4. `exists` - with either a negatable companion `notExists` or boolean argument
5. `prefix`, which takes a string
6. `excludes` (aka `anything-but`), which has 5 possible types as argument:
    1. a plain string
    2. a plain number
    3. an array of strings
    4. an array of numbers
    5. a string, which denotes a prefix
   If it makes it easier for jsii, we could create 5 `exclude` variants, such as: `excludeNumbers(1, 2, 3, 4)`, `excludePrefix("test-")` and so on

Reply:
Sounds good to me. I would add a `json` static method, that takes in `json: any` and acts as a catch-all to any users who simply want to supply their patterns as a json object. 

Two options for `excludes`: we can have it take an `any` and then parse the types ourselves in one function, or we can have two separate functions, an `excludes(x: string[])` and `excludes(x: number[])`. We don't need to worry about the non-plural context; we can parse that out ourselves.

`excludePrefix` is probably the right way to go for the last option. 

Finally, I think `filterPatterns` should be of type `Record<string, FilterPatternRule>`. Or something of the sort. Just calling out that we like to use `Record` for our objects.

I'm not certain about any of this, but these are just my preliminary thoughts. 

If you want, feel free to ping me or request my review, sometimes things get lost in the shuffle.

Reply:
Friendly ping @wtho. Let me know if this is something you're still interested in taking on. If not, I'm happy to help bring this home, as there's certainly value in supporting this.

Reply:
Hey yes, I am still open to go on with it, but I will be busy in the next week. If you have the capacity, you take it ahead.

I realized there is quite some work done in https://github.com/aws/aws-cdk/blob/master/packages/%40aws-cdk/aws-events/lib/event-pattern.ts already, but the definitions there are not compliant with the actual spec, `string[]` does simplify the patterns too much, so this should be reworked. E. g. this example [taken from the docs](https://docs.aws.amazon.com/AmazonCloudWatch/latest/events/CloudWatchEventsandEventPatterns.html#CloudWatchEventsPatterns-null-and-empty-strings) throws an type error, as `null` is not a `string`:
```ts
const rule = new events.Rule(this, 'Rule', {
  eventPattern: {
    id: [ null ]
  },
})
```
And there are much more complex cases.

I think making a fully spec-compliant, high-level working example for Lambda Event Sources which later can replace the other event pattern APIs is the way to go.
But it is just not as simple as I thought working with the constraints of creating a JSII-idiomatic approach.

If you want to dive into this, this rule combination should be the most complex one to describe:
```
FilterCriteria: {
  Filters: [
    {
      Pattern: '{"dynamodb":{"nullFIeld":[null],"emptyField":[""],"matchFIeld":["a-match","alternate-match"],"notMatchFIeld":[{"anything-but":["not-match"]}],"numericField":[{"numeric":[">",10,"<=",20]}],"existsField":[{"exists":true}],"beginsWithField":[{"prefix":"us-"}],"nested":{"nestedField":["nested"]}}}',
    },
    {
      Pattern: '{"dynamodb":{"secondPatternField":["second"]}}',
    },
  ],
}
```

Which applies the pattern rules in all its flavors according the [eventbridge spec](https://docs.aws.amazon.com/eventbridge/latest/userguide/eb-event-patterns.html):
* 2 Filter Patterns (officially there can be up to 5)
* `null` value
* empty string
* string
* not
* numeric
* exist
* begins with
* nested

I guess even combinations should be possible, although I have not seen an example in the docs:
`"region": [ { "prefix": "us-"}, { "anything-but": [ "us-east-1" ] } ]`

If the new API will be able to produce all these patterns in the CFN template, it should be feature complete.

Reply:
It might be worth starting in a new PR with just the proposed readme changes so that we can go over it one more time and also pull in some other people that might have comments. This is going to introduce a fairly large surface area, and my advice is the more iterations that we get on the proposed changes, the faster the review process will go.

Of course, I know that it's been a couple months since your last comment, if this is something you don't have the bandwidth for, please let me know!

Reply:
We should also look to the prior art of sns subscription filters: https://github.com/aws/aws-cdk/tree/master/packages/%40aws-cdk/aws-sns#filter-policy